File: 比赛系统/src/batch_reporting.py
from __future__ import annotations

from pathlib import Path

from schemas import MarketPidSnapshot, PatternResult, PredictResult


def build_batch_summary(
    trade_date: str,
    sample_count: int,
    output_count: int,
    warnings: list[str],
) -> dict:
    return {
        "trade_date": trade_date,
        "sample_count": sample_count,
        "output_count": output_count,
        "warning_count": len(warnings),
        "warnings": warnings,
    }


def build_performance_summary(
    *,
    profile_enabled: bool,
    total_seconds: float,
    sample_build_seconds: float,
    pattern_seconds: float,
    capital_seconds: float,
    market_seconds: float,
    export_seconds: float,
    sample_timings: list[dict[str, float | str]],
    processed_samples: int,
    imputed_predict_count: int,
    skipped_incomplete_samples: int,
    round_seconds,
) -> dict | None:
    if not profile_enabled:
        return None

    top_slowest_samples = sorted(
        sample_timings,
        key=lambda item: float(item["sample_build_seconds"]),
        reverse=True,
    )[:20]
    return {
        "total_seconds": round_seconds(total_seconds),
        "sample_build_seconds": round_seconds(sample_build_seconds),
        "pattern_seconds": round_seconds(pattern_seconds),
        "capital_seconds": round_seconds(capital_seconds),
        "market_seconds": round_seconds(market_seconds),
        "export_seconds": round_seconds(export_seconds),
        "processed_samples": processed_samples,
        "imputed_missing_symbols": imputed_predict_count,
        "skipped_incomplete_samples": skipped_incomplete_samples,
        "top_slowest_samples": top_slowest_samples,
    }


def build_batch_result(
    *,
    trade_date: str,
    sample_count: int,
    pattern_results: list[PatternResult],
    predict_results: list[PredictResult],
    market_snapshot: MarketPidSnapshot | None,
    market_snapshot_path: Path | None,
    market_report_path: Path | None,
    market_validation_report_path: str,
    replay_validation_report_path: str,
    diagnostics_json_path: str,
    distribution_csv_path: str,
    submit_zip: str | None,
    warnings: list[str],
    imputed_output_count: int,
    stock_offset: int,
    stock_limit: int | None,
    stock_list_file: str | Path | None,
    stock_universe_size: int | None,
    missing_symbols: list[str],
    incomplete_stock_dirs: dict[str, list[str]],
    performance_summary: dict | None,
) -> dict:
    return {
        "trade_date": trade_date,
        "pattern_results": pattern_results,
        "predict_results": predict_results,
        "market_pid_snapshot": market_snapshot,
        "market_snapshot_path": str(market_snapshot_path) if market_snapshot_path else None,
        "market_report_path": str(market_report_path) if market_report_path else None,
        "market_validation_report_path": market_validation_report_path,
        "replay_validation_report_path": replay_validation_report_path,
        "diagnostics_json_path": diagnostics_json_path,
        "distribution_csv_path": distribution_csv_path,
        "submit_zip": submit_zip,
        "warnings": warnings,
        "sample_count": sample_count,
        "imputed_output_count": imputed_output_count,
        "output_count": len(pattern_results),
        "stock_offset": stock_offset,
        "stock_limit": stock_limit,
        "stock_list_file": str(stock_list_file) if stock_list_file else None,
        "stock_universe_size": stock_universe_size,
        "missing_symbols": missing_symbols,
        "incomplete_stock_dirs": incomplete_stock_dirs,
        "performance_summary": performance_summary,
    }

File: src-rust/src/batch_reporting.rs
use crate::schemas::{MarketPidSnapshot, PatternResult, PredictResult};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::path::Path;

pub fn build_batch_summary(
    trade_date: &str,
    sample_count: usize,
    output_count: usize,
    warnings: &[String],
) -> serde_json::Map<String, Value> {
    let mut summary = serde_json::Map::new();
    summary.insert("trade_date".into(), Value::String(trade_date.to_string()));
    summary.insert("sample_count".into(), json!(sample_count));
    summary.insert("output_count".into(), json!(output_count));
    summary.insert("warning_count".into(), json!(warnings.len()));
    summary.insert(
        "warnings".into(),
        Value::Array(warnings.iter().cloned().map(Value::String).collect()),
    );
    summary
}

#[allow(clippy::too_many_arguments)]
pub fn build_performance_summary(
    profile_enabled: bool,
    total_seconds: f64,
    sample_build_seconds: f64,
    pattern_seconds: f64,
    capital_seconds: f64,
    market_seconds: f64,
    export_seconds: f64,
    sample_timings: &[HashMap<String, f64>],
    processed_samples: usize,
    imputed_predict_count: usize,
    skipped_incomplete_samples: usize,
    round_seconds: fn(f64) -> f64,
) -> Option<Value> {
    if !profile_enabled {
        return None;
    }
    let mut top_slowest_samples = sample_timings.to_vec();
    top_slowest_samples.sort_by(|left, right| {
        right
            .get("sample_build_seconds")
            .unwrap_or(&0.0)
            .partial_cmp(left.get("sample_build_seconds").unwrap_or(&0.0))
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    top_slowest_samples.truncate(20);
    let top_payload: Vec<Value> = top_slowest_samples
        .into_iter()
        .map(|item| {
            let mut payload = serde_json::Map::new();
            if let Some(value) = item.get("sample_build_seconds") {
                payload.insert("sample_build_seconds".into(), json!(round_seconds(*value)));
            }
            Value::Object(payload)
        })
        .collect();
    Some(json!({
        "total_seconds": round_seconds(total_seconds),
        "sample_build_seconds": round_seconds(sample_build_seconds),
        "pattern_seconds": round_seconds(pattern_seconds),
        "capital_seconds": round_seconds(capital_seconds),
        "market_seconds": round_seconds(market_seconds),
        "export_seconds": round_seconds(export_seconds),
        "processed_samples": processed_samples,
        "imputed_missing_symbols": imputed_predict_count,
        "skipped_incomplete_samples": skipped_incomplete_samples,
        "top_slowest_samples": top_payload,
    }))
}

#[allow(clippy::too_many_arguments)]
pub fn build_batch_result(
    trade_date: &str,
    sample_count: usize,
    pattern_results: &[PatternResult],
    predict_results: &[PredictResult],
    market_snapshot: Option<&MarketPidSnapshot>,
    market_snapshot_path: Option<&Path>,
    market_report_path: Option<&Path>,
    market_validation_report_path: &str,
    replay_validation_report_path: &str,
    diagnostics_json_path: &str,
    distribution_csv_path: &str,
    submit_zip: Option<&str>,
    warnings: &[String],
    imputed_output_count: usize,
    stock_offset: usize,
    stock_limit: Option<usize>,
    stock_list_file: Option<&Path>,
    stock_universe_size: Option<usize>,
    missing_symbols: &[String],
    incomplete_stock_dirs: Value,
    performance_summary: Option<Value>,
) -> HashMap<String, Value> {
    let mut result = HashMap::new();
    result.insert("trade_date".into(), Value::String(trade_date.to_string()));
    result.insert("sample_count".into(), json!(sample_count));
    result.insert("output_count".into(), json!(pattern_results.len()));
    result.insert("imputed_output_count".into(), json!(imputed_output_count));
    result.insert("stock_offset".into(), json!(stock_offset));
    result.insert(
        "stock_limit".into(),
        stock_limit.map(Value::from).unwrap_or(Value::Null),
    );
    result.insert(
        "stock_list_file".into(),
        stock_list_file
            .map(|path| Value::String(path.to_string_lossy().to_string()))
            .unwrap_or(Value::Null),
    );
    result.insert(
        "stock_universe_size".into(),
        stock_universe_size.map(Value::from).unwrap_or(Value::Null),
    );
    result.insert(
        "warnings".into(),
        Value::Array(warnings.iter().cloned().map(Value::String).collect()),
    );
    result.insert(
        "missing_symbols".into(),
        Value::Array(missing_symbols.iter().cloned().map(Value::String).collect()),
    );
    result.insert("incomplete_stock_dirs".into(), incomplete_stock_dirs);
    result.insert(
        "submit_zip".into(),
        submit_zip.map(|path| Value::String(path.to_string())).unwrap_or(Value::Null),
    );
    result.insert(
        "market_snapshot_path".into(),
        market_snapshot_path
            .map(|path| Value::String(path.to_string_lossy().to_string()))
            .unwrap_or(Value::Null),
    );
    result.insert(
        "market_report_path".into(),
        market_report_path
            .map(|path| Value::String(path.to_string_lossy().to_string()))
            .unwrap_or(Value::Null),
    );
    result.insert(
        "market_validation_report_path".into(),
        Value::String(market_validation_report_path.to_string()),
    );
    result.insert(
        "replay_validation_report_path".into(),
        Value::String(replay_validation_report_path.to_string()),
    );
    result.insert(
        "diagnostics_json_path".into(),
        Value::String(diagnostics_json_path.to_string()),
    );
    result.insert(
        "distribution_csv_path".into(),
        Value::String(distribution_csv_path.to_string()),
    );
    result.insert(
        "performance_summary".into(),
        performance_summary.unwrap_or(Value::Null),
    );
    result.insert(
        "market_pid_snapshot".into(),
        market_snapshot
            .map(|snapshot| json!({
                "trade_date": snapshot.trade_date,
                "market_regime": snapshot.market_regime,
                "up_count": snapshot.up_count,
                "down_count": snapshot.down_count,
            }))
            .unwrap_or(Value::Null),
    );
    result.insert("pattern_result_count".into(), json!(pattern_results.len()));
    result.insert("predict_result_count".into(), json!(predict_results.len()));
    result
}
